signature/jws/spec.go

```diff
@@ -4,46 +4,97 @@
 package jws
 
 import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"errors"
+	"fmt"
 	"time"
 
 	"github.com/golang-jwt/jwt/v4"
 	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/spec/v1/signature"
 )
 
-// unprotectedHeader contains the header parameters that are not integrity protected.
-type unprotectedHeader struct {
-	TimeStampToken []byte   `json:"timestamp,omitempty"`
-	CertChain      [][]byte `json:"x5c,omitempty"`
-}
-
-// MediaTypeNotationPayload describes the media type of the payload of notation signature.
-const MediaTypeNotationPayload = "application/vnd.cncf.notary.v2.jws.v1"
-
-// payload contains the subject manifest and other attributes that have to be integrity
-// protected.
-type payload struct {
-	Notation notationClaim `json:"notary"`
+type notaryClaim struct {
 	jwt.RegisteredClaims
-}
-
-// notationClaim is the top level node and private claim, encapsulating the notary v2 data.
-type notationClaim struct {
-	Subject notation.Descriptor `json:"subject"`
+	Notary signature.JWSNotaryClaim `json:"notary"`
 }
 
 // packPayload generates JWS payload according the signing content and options.
-func packPayload(desc notation.Descriptor, opts notation.SignOptions) *payload {
+func packPayload(desc signature.Descriptor, opts notation.SignOptions) jwt.Claims {
 	var expiresAt *jwt.NumericDate
 	if !opts.Expiry.IsZero() {
 		expiresAt = jwt.NewNumericDate(opts.Expiry)
 	}
-	return &payload{
-		Notation: notationClaim{
-			Subject: desc,
-		},
+	return notaryClaim{
 		RegisteredClaims: jwt.RegisteredClaims{
-			IssuedAt:  jwt.NewNumericDate(time.Now()),
 			ExpiresAt: expiresAt,
+			IssuedAt:  jwt.NewNumericDate(time.Now()),
 		},
+		Notary: signature.JWSNotaryClaim{Subject: desc},
+	}
+}
+
+func checkCertChain(certChain []*x509.Certificate) error {
+	if len(certChain) == 0 {
+		return nil
+	}
+	if err := verifyCert(certChain[0], x509.ExtKeyUsageCodeSigning); err != nil {
+		return fmt.Errorf("signing certificate does not meet the minimum requirements: %w", err)
+	}
+	for _, c := range certChain[1:] {
+		for _, ext := range c.ExtKeyUsage {
+			if ext == x509.ExtKeyUsageTimeStamping {
+				if err := verifyCert(c, x509.ExtKeyUsageTimeStamping); err != nil {
+					return fmt.Errorf("timestamping certificate does not meet the minimum requirements: %w", err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+var (
+	oidExtensionKeyUsage = []int{2, 5, 29, 15}
+)
+
+// validateCert checks cert meets the requirements defined in
+// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#certificate-requirements.
+func verifyCert(cert *x509.Certificate, extKeyUsage x509.ExtKeyUsage) error {
```

Comment:
The name `verifyCert` is misleading. Will `verifyCertExtKeyUsage` be better?

Reply:
Sure

Reply:
The certificate requirements can include KU, EKU, basicConstraints etc. I'd name it VerifyCertRequirements or similar.

Reply:
I'd defer any further changes to this method till notaryproject/notaryproject#157 is resolved.